rend3-routine/src/skinning/mod.rs

```diff
@@ -0,0 +1,272 @@
+    pub skinning_inputs_bgl: BindGroupLayout,
+}
+
+impl GpuSkinner {
+    const WORKGROUP_SIZE: u32 = 64;
+
+    pub fn new(device: &wgpu::Device) -> GpuSkinner {
+        let storage_buffer_ty = |read_only| wgpu::BindingType::Buffer {
+            ty: wgpu::BufferBindingType::Storage { read_only },
+            has_dynamic_offset: false,
+            min_binding_size: None,
+        };
+
+        // Bind group 0 contains some vertex buffers bound as storage buffers
+        let vertex_buffers_bgl = {
+            let mut bglb = BindGroupLayoutBuilder::new();
+            bglb.append(wgpu::ShaderStages::COMPUTE, storage_buffer_ty(false), None); // Positions
+            bglb.append(wgpu::ShaderStages::COMPUTE, storage_buffer_ty(false), None); // Normals
+            bglb.append(wgpu::ShaderStages::COMPUTE, storage_buffer_ty(false), None); // Tangents
+            bglb.append(wgpu::ShaderStages::COMPUTE, storage_buffer_ty(false), None); // Joint indices
+            bglb.append(wgpu::ShaderStages::COMPUTE, storage_buffer_ty(false), None); // Joint weights
+            bglb.build(device, Some("Gpu skinning vertex buffers"))
+        };
+
+        // Bind group 1 contains the joint matrices global buffer. Each
+        // skinning input struct contains an offset into this array.
+        let joint_matrices_bgl = {
+            let mut bglb = BindGroupLayoutBuilder::new();
+            bglb.append(wgpu::ShaderStages::COMPUTE, storage_buffer_ty(true), None);
+            bglb.build(device, Some("Gpu skinning joint matrices"))
+        };
+
+        // Bind group 2 contains the pre skinning inputs. This uses dynamic
+        // offsets because there is one dispatch per input, and the offset is
+        // used to indicate which is the current input to the shader.
+        //
+        // NOTE: This would be an ideal use case for push constants, but they are
+        // not available on all platforms so we need to use this workaround.
+        let skinning_inputs_bgl = {
+            let mut bglb = BindGroupLayoutBuilder::new();
+            bglb.append(
+                wgpu::ShaderStages::COMPUTE,
+                wgpu::BindingType::Buffer {
+                    ty: wgpu::BufferBindingType::Uniform,
+                    has_dynamic_offset: true,
+                    min_binding_size: None,
+                },
+                None,
+            );
+            bglb.build(device, Some("Gpu skinning inputs"))
+        };
+
+        let layout = device.create_pipeline_layout(&wgpu::PipelineLayoutDescriptor {
+            label: None,
+            bind_group_layouts: &[&vertex_buffers_bgl, &joint_matrices_bgl, &skinning_inputs_bgl],
+            push_constant_ranges: &[],
+        });
+
+        let module = device.create_shader_module(&wgpu::ShaderModuleDescriptor {
+            label: Some("Gpu skinning compute shader"),
+            source: wgpu::ShaderSource::Wgsl(include_str!("../../shaders/src/skinning.wgsl").into()),
+        });
```

Comment:
I'm not sure how the shader loading mechanism in rend3-routine works, and to be honest I'm also not very familiar with the specific flavor of GLSL that Vulkan/Wgpu use. So I went for a wgsl shader instead.

Reply:
This kind of steady transition is what's going to work well to transition, so I'm glad you're starting it!